src/libraries/System.Net.Connections/src/System/Net/Connections/ConnectionExtensions.cs

```diff
@@ -0,0 +1,266 @@
+            if (filter == null) throw new ArgumentNullException(nameof(filter));
+            return new ConnectionFilteringFactory(factory, filter);
+        }
+
+        /// <summary>
+        /// Creates a connection-level filter on top of a <see cref="ConnectionFactory"/>.
+        /// </summary>
+        /// <param name="factory">The factory to be filtered.</param>
+        /// <param name="filter">The connection-level filter to apply on top of <paramref name="factory"/>.</param>
+        /// <returns>A new filtered <see cref="ConnectionFactory"/>.</returns>
+        public static ConnectionFactory Filter(this ConnectionFactory factory, Func<Connection, IConnectionProperties?, CancellationToken, ValueTask<Stream>> filter)
+        {
+            if (factory == null) throw new ArgumentNullException(nameof(factory));
+            if (filter == null) throw new ArgumentNullException(nameof(filter));
+            return new StreamFilteringFactory(factory, filter);
+        }
+
+        private sealed class ConnectionFilteringFactory : ConnectionFactory
+        {
+            private readonly ConnectionFactory _baseFactory;
+            private readonly Func<Connection, IConnectionProperties?, CancellationToken, ValueTask<Connection>> _filter;
+
+            public ConnectionFilteringFactory(ConnectionFactory baseFactory, Func<Connection, IConnectionProperties?, CancellationToken, ValueTask<Connection>> filter)
+            {
+                _baseFactory = baseFactory;
+                _filter = filter;
+            }
+
+            public override async ValueTask<Connection> ConnectAsync(EndPoint? endPoint, IConnectionProperties? options, CancellationToken cancellationToken = default)
+            {
+                Connection con = await _baseFactory.ConnectAsync(endPoint, options, cancellationToken).ConfigureAwait(false);
+                try
+                {
+                    return await _filter(con, options, cancellationToken).ConfigureAwait(false);
+                }
+                catch
+                {
+                    await con.DisposeAsync().ConfigureAwait(false);
+                    throw;
+                }
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                if (disposing) _baseFactory.Dispose();
+            }
+
+            protected override ValueTask DisposeAsyncCore()
+            {
+                return _baseFactory.DisposeAsync();
+            }
+        }
+
+        private sealed class StreamFilteringFactory : ConnectionFactory
+        {
+            private readonly ConnectionFactory _baseFactory;
+            private readonly Func<Connection, IConnectionProperties?, CancellationToken, ValueTask<Stream>> _filter;
+
+            public StreamFilteringFactory(ConnectionFactory baseFactory, Func<Connection, IConnectionProperties?, CancellationToken, ValueTask<Stream>> filter)
+            {
```

Comment:
Why does it take CancellationToken if it is not used?

Reply:
It takes a `Func` that takes a `CancellationToken`.